agent/hcp/telemetry/exporter.go

```diff
@@ -0,0 +1,169 @@
+package telemetry
+
+import (
+	"context"
+	"fmt"
+	"regexp"
+
+	"github.com/armon/go-metrics"
+	hcpclient "github.com/hashicorp/consul/agent/hcp/client"
+	"github.com/hashicorp/go-hclog"
+	"go.opentelemetry.io/otel/attribute"
+	"go.opentelemetry.io/otel/sdk/instrumentation"
+	"go.opentelemetry.io/otel/sdk/metric"
+	"go.opentelemetry.io/otel/sdk/metric/metricdata"
+	"go.opentelemetry.io/otel/sdk/resource"
+)
+
+type Exporter interface {
+	Export(ctx context.Context, goMetrics []*metrics.IntervalMetrics) error
+	ConvertToOTLP(goMetrics []*metrics.IntervalMetrics) []metricdata.Metrics
+}
+
+type MetricsExporterConfig struct {
+	Labels  map[string]string
+	Logger  hclog.Logger
+	Filters []string
+	Client  hcpclient.Client
+}
+
+type MetricsExporter struct {
+	client   hcpclient.Client
+	resource *resource.Resource
+	logger   hclog.Logger
+	filter   *FilterList
+}
+
+// NewMetricsExporter returns a Metrics exporter with labels to tag metrics
+// and filters to only export allowed metrics.
+func NewMetricsExporter(cfg *MetricsExporterConfig) (*MetricsExporter, error) {
+	if cfg.Client == nil || cfg.Logger == nil {
+		return nil, fmt.Errorf("HCP client and a logger are required")
+	}
+
+	attr := make([]attribute.KeyValue, 0, len(cfg.Labels))
+	for key, val := range cfg.Labels {
+		attr = append(attr, attribute.KeyValue{
+			Key:   attribute.Key(key),
+			Value: attribute.StringValue(val),
+		})
+	}
+
+	f := &FilterList{map[string]*regexp.Regexp{}}
+	if err := f.Set(cfg.Filters); err != nil {
+		return nil, fmt.Errorf("invalid regex: %v", err.Error())
```

Comment:
TODO: When we have bad filters, I think we should just log the error instead of failing the metrics collection. Will update this but if anyone has any thoughts please let me know!